Packages/Tracking/Hands/Runtime/Scripts/HandModelManager.cs

```diff
@@ -0,0 +1,512 @@
+            }
+            else
+            {
+                handModelPairs[handModelPairIndex].Right = handModel;
+                handModelPairs[handModelPairIndex].RightEnableDisable = handEnableDisable;
+            }
+
+            if (handModelPairs[handModelPairIndex].HandModelPairId == null || handModelPairs[handModelPairIndex].HandModelPairId == "")
+            {
+                handModelPairs[handModelPairIndex].HandModelPairId = GetDistinctHandModelId(handModel.gameObject.transform.parent.name);
+            }
+        }
+
+        protected string GetDistinctHandModelId(string desiredHandModelId)
+        {
+            int iteration = 0;
+            var name = desiredHandModelId;
+            List<string> existingIds = handModelPairs.Select(hmp => hmp.HandModelPairId).ToList();
+
+            while (existingIds.Contains(name))
+            {
+                name = desiredHandModelId + "_[" + (iteration++) + "]";
+            }
+
+            return name;
+        }
+
+        /// <summary>
+        /// Generates a hand model index to register with the hand model list
+        /// If attempting to find a pair & the pair is already registered, uses that pair's index
+        /// Otherwise assigns the next index in the list
+        /// </summary>
+        /// <param name="handModel">the hand model to register</param>
+        /// <param name="attemptToFindPair">if true, attempts to find a pair & register it if unregistered</param>
+        /// <returns>The correct hand model index</returns>
+        private int GenerateHandModelIndex(HandModelBase handModel, bool attemptToFindPair)
+        {
+            int handModelPairIndex = handModelPairs.Count;
+
+            if (!attemptToFindPair)
+            {
+                handModelPairs.Add(new HandModelPair());
+            }
+
+            // Look for hand model pair in previously registered hands
+            for (int i = 0; i < handModelPairs.Count; i++)
+            {
+                HandModelBase potentialRegisteredHandPair;
+
+                if (handModel.Handedness == Chirality.Left)
+                {
+                    potentialRegisteredHandPair = handModelPairs[i].Right;
+                }
+                else
+                {
+                    potentialRegisteredHandPair = handModelPairs[i].Left;
+                }
+
+                if (handModel.gameObject.transform.parent == potentialRegisteredHandPair.gameObject.transform.parent)
+                {
+                    handModelPairIndex = i;
+                    return handModelPairIndex;
+                }
+            }
+
+            //If not found in previously registered hands, look for in the parent transform
+            HandModelBase potentialHandPair = handModel.gameObject.transform.parent.GetComponentsInChildren<HandModelBase>().FirstOrDefault(hmb => hmb != handModel);
+
+            handModelPairs.Add(new HandModelPair());
```

Comment:
Why are we doing this potentially twice?

Reply:
fixed :)